### ISSUE TITLE
Custom joins with Sequelize.literal

### ISSUE THREAD

**Issue by gavinbelson** [sequelize/sequelize#11788]
## Issue Description
When you need to add some custom table or a join for a report, there is not always an association defined in the models. For this other ORMs like Rails' active record allow you to define custom joins without setting up an association.

## Example
`models.user.findAll({
    include: [
      {model: models.user_report, as: 'reports', on: Sequelize.literal('user.id = reports.user_id')}
    ] ...`

For this to work I have to have a user to user_report association. This is not always possible and inconvenient to always define. 

### Describe the solution you'd like
The easiest this would be to allow a completely custom Sequelize.literal() inside the include:

`include[ {Sequelize.literal( ' INNER JOIN user_report reports ON user.id = reports.user_id' )} ]`

### Why should this be in Sequelize
This is how it is done with other ORMs. As more users switch from Rails, Django and Laravel to node, they will be looking for something like this. There are several Stack Overflow questions asking for this. Ex: https://stackoverflow.com/questions/47215961/sequelize-js-join-tables-without-associations

Many issues brought up in the sequelize github could also be easier resolved with this as opposed to looking for all the operators etc within the include clause. Ex: https://github.com/sequelize/sequelize/issues/6718

### Describe alternatives/workarounds you've considered
The only way I see to not have an association and still joining is a raw query, which erases the convenience of the ORM

### Is this issue dialect-specific?

- [X] No. This issue is relevant to Sequelize as a whole.
- [ ] Yes. This issue only applies to the following dialect(s): XXX, YYY, ZZZ

### Would you be willing to resolve this issue by submitting a Pull Request?

- [ ] Yes, I have the time and I know how to start.
- [ ] Yes, I have the time but I don't know how to start, I would need guidance.
- [X] No, I don't have the time, although I believe I could do it if I had the time...
- [ ] No, I don't have the time and I wouldn't even know how to start.

**Comment by gavinbelson** [comment]
This may be really simple to implement.. just add a "joins" parameter which the can be filled out with the Sequelize.literal() On the backend, anything in the "joins" parameter would just get appended to the end of the JOIN part of the SQL query that sequelize creates

**Comment by papb** [comment]
Hello, thanks for the issue report. Can you please expand on your example:

* Providing the full SQL you are trying to obtain with Sequelize, without having to define/use an association explicitly
* Show more explicitly how you currently do it by defining an association explicitly

Thanks :grimacing:

**Comment by gavinbelson** [comment]
Sure, thanks for looking at it:

SQL:
`SELECT reports.preferences, user.name FROM users
 INNER JOIN user_report reports ON user.id = reports.user_id`

Currently, I have to define this association in the user model, then I can get the sequelize include to work:
  user.associate = function(models) {
    user.hasMany(models.user_report);
  };

The point of this would be to give more flexibility to use a literal to define a join, this would be widely used. In my case, not every table/view needs a relationship and I could use a literal to join those tables.

Let me know if you have any other questions

**Comment by gavinbelson** [comment]
Hey guys, any update here? Can I help?

**Comment by jiangguoliang563** [comment]
I have the same problem

**Comment by jiangguoliang563** [comment]
I have the same problem, want to use 'left join' without define association

**Comment by papb** [comment]
Can you please take a look at the [new guide on sub queries](https://sequelize.org/master/manual/sub-queries.html) and let me know if that helps?

**Comment by gavinbelson** [comment]
That's close, but a subquery creates a whole new query inside the original query.

We just need to be able to define the literal `JOIN` to add onto the original query.  

For example, we just need to be able to add this literal SQL to the query:
`INNER JOIN user_report reports ON user.id = reports.user_id`

**Comment by papb** [comment]
@gavinbelson Ah, ok. Makes sense.

**Comment by papb** [comment]
> Can I help?

@gavinbelson That would be great. Would you like to send a PR?

**Comment by gavinbelson** [comment]
@papb yes but need hand-holding..  what would the process be? Clone the git repo, then how do you test the additions - aka where do you guys usually write the sequelize query to test your work?

**Comment by papb** [comment]
When I want to play around with some code I like to use my [sequelize-sscce](https://github.com/papb/sequelize-sscce) repository... I plan on adding a nice way to fiddle with sequelize internals directly in that repo in the near future. For now what I suggest you to do is to clone the [sequelize-sscce](https://github.com/papb/sequelize-sscce) repository, play around with it in debug mode (works out-of-the-box if you're using VSCode for example) and to test ideas you edit the code inside node_modules directly (a bit ugly but I will improve this in the near future).

This is how I do it, not sure about the other maintainers... Let me know if you have any trouble :)

**Comment by gavinbelson** [comment]
Thanks, I'll give it a shot. 

Looking at the lib right now, where do you guys store the include statement logic and sequelize literal logic?

**Comment by papb** [comment]
@gavinbelson Sorry, I had to leave. Did you figure it out?

**Comment by Solid-Metal** [comment]
hey just chime in, is there any update regarding this?, the reason why i want to use literal join is because for some reason sequelize not allow me to use outer join, even tho i already googled and it should work just fine

**Comment by cedgut** [comment]
Any update on this? A literal join would really make things a lot smoother to work with.

**Comment by Vaelek** [comment]
Bump.

**Comment by ajmcgarry1608** [comment]
This is really important for another reason.
A WHERE clause condition using a field from the LEFT/RIGHT joined table will always nullify the effect of the left join because WHERE acts like a filter post join. Being able to add an arbitrary cross table condition to a LEFT/RIGHT JOIN means you can conditionally join on this condition rather than just filter.
Completely different results

**Comment by mopcweb** [comment]
Hi!
Are there any updates?
This one is really very important functionality ...

**Comment by gavinbelson** [comment]
I never got the questions answered about where the `where` logic lies and unfortunately don't have time right now to put this in. We need a contributor. @felixfbecker know anyone that can put this in?

**Comment by gavinbelson** [comment]
Does anyone know a project maintainer? This is a key issue for sequelize popularity

**Comment by shakedlokits** [comment]
This is a key issue for us as well, is there a resource/planning on this feature?

**Comment by thalesfsp** [comment]
Any update on this?

**Comment by ephys** [comment]
Still the same status, nobody from the core team is currently working on this.
Someone from the community taking the lead on this is the current best chance to get this implemented until either someone from the team takes special interest in implementing this, or we manage to free our backlog enough to get to this (which you can promote by reacting with 👍🏼 on the initial post)

**Comment by normandanielmq** [comment]
Just now I need to add a custom SQL in the form clause
```
select s.id,program_name as program_id
from scholarships s, 
jsonb_array_elements((s.json#>'{ScholarshipProgramCollection,ScholarshipProgramInfo, EligibilityCriteria}')) as ec
where ec#>'{Academics}' @? '$[*] ? (@.AcademicEligibility == "Minimum GPA")'
```

**Comment by thomas-p-wilson** [comment]
This has worked for me on a number of occasions:

```
{
  association: {
    as: 'task',
    foreignKey: 'task_id',
    targetKey: 'id',
  },
  on: literal('"System"."task_id" = "task"."id"),
  attributes: ['id', 'name'],
}
```

A somewhat contrived example, so apologies for that. But the important point here is mimicking the structure of the association.

I can't take credit for this. I found it somewhere else. But I cannot for the life of me remember where, as it has been well over a year now since I found that.

**Comment by thomas-p-wilson** [comment]
Working example from my TS codebase:

```
{
  association: {
    as: 'task',
    foreignKey: 'table_row_id',
    targetKey: 'id',
  },
  on: literal('"System".system_template_id IS NULL AND LOWER(recommendation.name) = LOWER("System".name) AND (recommendation.organization_id = 0 OR recommendation.organization_id IS null)'),
  attributes: ['id', 'name'],
  required: false,
}
```

**Comment by SiarheiBokuts** [comment]
Hi! I totally agree that it would be great to have the ability to add custom code inside included models.

For example, something like:

```
include: [
  {
    Sequelize.literal('INNER JOIN ...')
  }
]
```

I spent a lot of time trying to find a solution for this, but I couldn't find anything that worked. It seems that this option isn't available at the moment, and the only workaround I've found is to use two separate queries.

**Comment by patricktyndall** [comment]
Just chiming in that this would be extremely valuable. We found a way to do this, but it feels fragile and would love better support. This has repeatedly proven useful to our team to expose relationships between high level models when the underlying relationship is more complex in DB.

Linking my other comment [here](https://github.com/sequelize/sequelize/issues/5385#issuecomment-3014439657) with more info.
- Trickiest things: A) figuring out query aliasing (e.g. if I want to "chain" these into a nested `includes`, i.e. FK relationship -> "custom" relationship -> FK relationship), and B) the fact that `foreignKey` is necessary to avoid crash but is not used by the library when `on:` is provided (as noted by @thomas-p-wilson ).

In addition to need for more official support, a key limitation is the inability to use **`separate: true`** (doesn't work when `on:` is used), which would be key for limiting data size high-cardinality custom relationships.

Thanks!

**Comment by Senkai350** [comment]
Is include.on now works if i try to include m2m relationship through junction table? I tried to add "on" in include statement but sequelize simply ignore it. Also through options object doesn't have 'on' property at all.
